### DISCUSSION TITLE
Difference between default_properties = "fips=no" & default_properties = "provider!=fips"

### DISCUSSION THREAD

**Discussion by sshedi** [openssl/openssl#22735] [Q&A] [answered]
Hi,

I am having a hard time understanding the difference between: (this is in openssl-3.x).

```
default_properties = "fips=no"
```
And

```
default_properties = "provider!=fips"
```

When "fips=no" `openssl md5 <file>` works fine.
But this command `openssl x509 -in certificate.pem -text -noout` fails with:
```
Could not open file or uri for loading certificate from certificate.pem
C0719377527F0000:error:16000069:STORE routines:ossl_store_get0_loader_int:unregistered scheme:crypto/store/store_register.c:237:scheme=file
C0719377527F0000:error:1608010C:STORE routines:inner_loader_fetch:unsupported:crypto/store/store_meth.c:383:No store loader found. For standard store loaders you need at least one of the default or base providers available. Did you forget to load them? Info: Global default library context, Scheme (file : 0), Properties (<null>)
Unable to load certificate
```

But using `default_properties = "provider!=fips"` decodes pem file properly and md5 also works fine.

So, what is the difference between these two? Please provide some more info on this.

My second question is, what is the proper way to disable openssl fips in openssl3.x using a conf file?

PS:
`default_properties = "-fips"` also works fine with md5 and pem decoding does that mean "provider!=fips" & "-fips" are same?

**Comment by t8m** [comment, marked as answer]
`fips=no` is bad query because non-crypto provided implementations such as stores, encoders and decoders can have `fips=yes` even if included in the default provider. So that's why `openssl x509` fails this way.

The best query to enable non-fips stuff is `default_properties = "-fips"` or simply empty default properties.

**Reply by sshedi** [reply]
Thanks a lot @t8m. Your explanation makes sense.

But the options provided by user in command line of through conf files should take precedence and behave accordingly.

My concern here is, if user says `fips=no` in conf file it should be considered a global setting and any implementations which do `fips=yes` internally should be overridden with user settings.

Please correct me if I'm wrong here.
